### PR TITLE
[lld][ELF] Use the containing symbol, not the referenced symbol, for undefined symbol errors.

### PR DESCRIPTION
**Pull request by bevin-hansson**
When reporting undefined symbols, the symbol being used to find
the source location for where the symbol was referenced from was
the symbol being referenced, not the symbol where the reference
was made from.

Use the relocation offset to locate the correct symbol.

With this, the following program:

```
extern int ex;

int * p[] = { &ex };

int main() {
  return 0;
}
```

When built with `clang -fuse-ld=lld a.c -g`, it goes from
```
ld.lld: error: undefined symbol: ex
>>> referenced by a.c
>>>               /tmp/a-0e7dc8.o:(p)
```
to
```
ld.lld: error: undefined symbol: ex
>>> referenced by a.c:3 ([...]/a.c:3)
>>>               /tmp/a-0e7dc8.o:(p)
```

I'm unsure how to construct a reasonable lit test for this.

### CODE REVIEW

File: lld/ELF/Relocations.cpp [hunk 1]
```diff
@@ -693,6 +693,34 @@ static const Symbol *getAlternativeSpelling(const Undefined &sym,
   return nullptr;
 }
 
+static Symbol &
+getSymAtOffset(InputSectionBase &sec, uint64_t off, Symbol &sym) {
```
Comment (smithp35) [inline lld/ELF/Relocations.cpp:697]: Would getSymContainingOffset() be a better name, as we're looking for the symbol with off in the range [sym->off,, sym->off + sym->size) .

**Review by smithp35 — COMMENTED** [review]
The code changes look good to me. I think it should be possible to write a lit test for this using the .file and .loc directives. The only example in the lld testsuite that does this is debug-line-obj.s 

You may be able to adapt that, or something similar to make a test case.

**Comment by MaskRay** [comment]
Thanks for the improvement. Can you change the description to mention https://reviews.llvm.org/D31481 ?

**Comment by MaskRay** [comment]
The `sym` parameter of `getSrcMsg` provides a fallback when line table information is unavailable.
The line table information covers cases when the undefined symbol is from a function.

> `int * p[] = { &ex };`

This is about a case when the undefined symbol is from a variable with no line table information.
I am nervous if we need to spend these lines of code to give a diagnostic for this case and I'd probably choose not.

Fortunately, I think we can refactor the existing code (`getEnclosingFunction`) to give this diagnostic improvement with nearly no additional complexity: 
https://github.com/MaskRay/llvm-project/tree/lld-enclosing-function I haven't added a test yet.

**Comment by bevin-hansson** [comment]
> I am nervous if we need to spend these lines of code to give a diagnostic for this case and I'd probably choose not.
> 
> Fortunately, I think we can refactor the existing code (`getEnclosingFunction`) to give this diagnostic improvement with nearly no additional complexity: https://github.com/MaskRay/llvm-project/tree/lld-enclosing-function I haven't added a test yet.

The change in your patch seems reasonable, since we can reuse the existing function for finding the enclosing symbol. Do you want me to transplant those changes into this PR?

**Comment by bevin-hansson** [comment]
Superseded by https://github.com/llvm/llvm-project/pull/70854

**Pull request closed without merging**